src/levanter/models/gemma.py

```diff
@@ -0,0 +1,608 @@
+        new_weights = hax.tree_util.resize_axis(self.token_embeddings, self.Vocab, new_size, key=key)
+        return dataclasses.replace(self, Vocab=self.Vocab.resize(new_size), token_embeddings=new_weights)
+
+
+class GemmaLMHeadModel(eqx.Module, LmHeadModel[GemmaConfig], StateDictSerializationMixin):
+    transformer: GemmaTransformer
+    embeddings: GemmaEmbedding
+    lm_head: hnn.Linear
+
+    @property
+    def config(self):
+        return self.transformer.config
+
+    @property
+    def vocab_size(self) -> int:
+        return self.Vocab.size
+
+    @property
+    def Vocab(self) -> Axis:
+        return self.embeddings.Vocab
+
+    @classmethod
+    def init(cls, Vocab: Axis, config: GemmaConfig, *, key) -> "GemmaLMHeadModel":
+        k_t, k_emb = jrandom.split(key, 2)
+        transformer = GemmaTransformer.init(config, key=k_t)
+        embeddings = GemmaEmbedding.init(Vocab, config, key=k_emb)
+        lm_head = hnn.Linear.init(In=config.Embed, Out=Vocab, key=k_emb, use_bias=False, out_first=True)
+        return GemmaLMHeadModel(transformer, embeddings, lm_head)
+
+    def __call__(
+        self,
+        input_ids: NamedArray,
+        attn_mask: Optional[Union[NamedArray, AttentionMask]] = None,
+        *,
+        key=None,
+    ) -> NamedArray:
+        """
+        Args:
+            input_ids (NamedArray): [batch, position]
+                Indices of input sequence tokens in the vocabulary.
+            attn_mask (Union[NamedArray, AttentionMask], optional): [batch, position]
+                Mask to avoid performing attention on the padding token indices of the encoder input.
+                The attn_mask from training pipeline may be an AttentionMask object instead of NamedArray
+        """
+        k_t, k_head = maybe_rng_split(key, 2)
+        x = self.embeddings.embed(input_ids)
+        x = self.transformer(x, attn_mask=attn_mask, key=k_t)
+        lm_logits = self.lm_head(x, key=k_head)
+        return lm_logits
+
+    def resize_vocab(self, new_size: int, key=None) -> "LmHeadModel[GemmaConfig]":
+        new_Vocab = self.Vocab.resize(new_size)
+        k1, k2 = maybe_rng_split(key, 2)
+        new_embeddings = self.embeddings.resize_embeddings(new_size, key=k1)
+        new_lm_matrix = hax.tree_util.resize_axis(self.lm_head.weight, self.Vocab, new_size, key=k2)
+        new_lm_head = dataclasses.replace(self.lm_head, Out=new_Vocab, weight=new_lm_matrix)
+
+        return dataclasses.replace(self, embeddings=new_embeddings, lm_head=new_lm_head)
+
+    def _state_dict_key_map(self) -> Dict[str, Optional[str]]:
+        """Map from Levanter model names to HF."""
+        return {"transformer": "model", "embeddings": None}
+
+    def from_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None):
+        # unflatten the linear layers of HF state_dict to match the shape of GemmaMlp
+        d = state_dict.copy()
+
+        # Gemma lm_head weights are tied (shared) with the embedding layer.
```

Comment:
yeah weight tying is something that isn't handled super well in Equinox (and thus Haliax) I typically just hard code the relationship if they're tied (gpt2 is implemented this way). Equinox's solution is [this](https://docs.kidger.site/equinox/api/nn/shared/) but I haven't really used it in anger yet.

Reply:
(Fine to leave it as is, or you can switch to the gpt2-style tying, or you can try the Equinox thing if you want. I'll leave it to you)

Reply:
Thanks for the pointeres! I'll take a look at these tomorrow (out today). I think we should do it, otherwise the output model won't be compatible with the HF Gemma anymore IIUC.

Reply:
Thanks for the tip, I implemented the gpt2-style tying as you mentioned and it worked perfectly.